packages/jest-runtime/src/transform.js

```diff
@@ -110,94 +116,189 @@ const readCacheFile = (filePath: Path, cachePath: Path): ?string => {
+  // Create sub folders based on the cacheKey to avoid creating one
+  // directory with many files.
+  const cacheDir = path.join(baseCacheDir, cacheKey[0] + cacheKey[1]);
+  const cachePath = path.join(
+    cacheDir,
+    path.basename(filename, path.extname(filename)) + '_' + cacheKey,
+  );
+  createDirectory(cacheDir);
+
+  return cachePath;
+};
+
+const preprocessorCache = new WeakMap();
+
+const getPreprocessor = (config: Config): ?Processor => {
+  if (preprocessorCache.has(config)) {
+    return preprocessorCache.get(config);
+  } else {
+    let preprocessor;
+    if (!config.scriptPreprocessor) {
+      preprocessor = null;
     } else {
-      content = preprocessor.process(content, filename, config);
-      if (options && options.instrument) {
-        content = options.instrument(content);
+      // $FlowFixMe
+      preprocessor = require(config.scriptPreprocessor);
+      if (typeof preprocessor.process !== 'function') {
+        throw new TypeError(
+          'Jest: a preprocessor must export a `process` function.',
+        );
       }
-      content = wrap(content);
-      writeCacheFile(cachePath, content);
     }
-  } else if (options && options.instrument) {
-    content = wrap(options.instrument(content));
+    preprocessorCache.set(config, preprocessor);
+    return preprocessor;
+  }
+};
+
+const stripShebang = content => {
+  // If the file data starts with a shebang remove it. Leaves the empty line
+  // to keep stack trace line numbers correct.
+  if (content.startsWith('#!')) {
+    return content.replace(/^#!.*/, '');
+  } else {
+    return content;
+  }
+};
+
+const instrument = (content: string, filename: Path): string => {
+  // NOTE: Keeping these requires inside this function reduces a single run
+  // time by 2sec if not running in `--coverage` mode
+  const babel = require('babel-core');
+  const babelPluginIstanbul = require('babel-plugin-istanbul').default;
+
+  if (babel.util.canCompile(filename)) {
+    return babel.transform(content, {
+      filename,
+      plugins: [babelPluginIstanbul],
+      retainLines: true,
+      babelrc: false,
+    }).code;
+  }
+  return content;
+};
+
+const cachedTransformAndWrap = (
+  filename: Path,
+  config: Config,
+  content: string
+) => {
+  const preprocessor = getPreprocessor(config);
+  const cacheFilePath = getFileCachePath(filename, config, content);
+  // Ignore cache if `config.cache` is set (--no-cache)
+  let result = config.cache ? readCacheFile(filename, cacheFilePath) : null;
+
+  if (result) {
+    return result;
+  }
+
+  result = content;
+
+  if (preprocessor && shouldPreprocess(filename, config)) {
+    result = preprocessor.process(result, filename, config);
+  }
+  if (shouldInstrument(filename, config)) {
+    result = instrument(result, filename, config);
+  }
+
+  result = wrap(result);
+  writeCacheFile(cacheFilePath, result);
+  return result;
+};
+
+const transformAndBuildScript = (
```

Comment:
i purposefully renamed it from `transform` to `transfomAndBuildScript` because it became to complicated.
This logic is pretty complex and it hard to read stuff like
`result = transform(...)`
because it can return multiple values (transformed, transformed and wrapped, transformed and wrapped and addet to vm.Script)
when i started switching back and forth between fn declaration and fn calls i renamed them to a more verbose version.

same for `cachedTransformAndWrap`. it's just much easier to read it